### DISCUSSION TITLE
[stm32h723zg] Uart interrupts get disabled while writing to internal flash

### DISCUSSION THREAD

**Discussion by sobkulir** [zephyrproject-rtos/zephyr#59714] [Q&A] [answered]
Hello,

I'm trying to send a file over USART2 on stm32h732zg and store it on the internal flash. However, while writing the file, interrupts get disabled for ~800ms so I can't receive more contents of the file during this period.

## Why??
Can you please help me figure out why this is happening?

1) There are only 4 free blocks for LittleFS which might land a hand for some weird behavior. This is because the sector size of the internal flash is 128kB and my partition is 512kB big. I don't know why the blocks can't be smaller than sectors, but I'm guessing this is a typical Filesystem limitation (clusters in FAT also consist of sectors).

2) The code is executed from the internal flash which is being written to. Could this pose a problem? I don't have experience with "executing from flash", all OSes I've worked with executed from RAM (not embedded). So I could imagine there being congestion for the flash device.

## Setup
To demonstrate the issue, I'm writing `0,1,2,...255` repeatedly over USART2 from a Python script, while on the MCU I check if the bytes are arriving in sequence. Then I create a 512kB file in two writes into LittleFS a partition. However, during the second write, the interrupts from USART2 stop coming for ~800ms. I attached my `main.cpp` and `nucleo_h723zg.overlay` and `prj.conf`: [app.zip](https://github.com/zephyrproject-rtos/zephyr/files/11850376/app.zip)

### Snippet
main.cpp
```cpp

void usart_interrupt_cb(const struct device *dev, void *user_data) {
  int *cur = reinterpret_cast<int *>(user_data);
  
  if (!uart_irq_update(dev)) {
    return;
  }

  if (!uart_irq_rx_ready(dev)) {
    return;
  }
  
  uint8_t c;
  while (uart_fifo_read(dev, &c, 1) == 1) {
    printk("%d (cur) %d\n", c, *cur);
         if (*cur != -1 && static_cast<int>(c) != (*cur + 1)%256) {
      printk("overrun, got: %d exp: %d\n", c, (*cur + 1)%256);
    }
    *cur = c;
  }
}

int main()
{
  // Mount LittleFS partition
  int rc;
  rc = littlefs_mount(mp);
  if (rc < 0) {
    return 1;
  }

  // Setup UART
  setup_usart();
    
  // Write file
  printk("Writing file\n");
  struct fs_file_t file;
  fs_file_t_init(&file);
  rc = fs_open(&file, "/seq/test.cpp", FS_O_WRITE | FS_O_CREATE);
  if (rc < 0) {
    printk("couldn't open file: %d\n", rc);
    return 1;
  }

  char buffer[256];
  // 512kB
  for (int i = 0 ; i < 2; ++i) {
    rc = fs_write(&file, buffer, sizeof(buffer));
    if (rc < 0) {
      printk("couldn't print file: %d\n", rc);
      return 1;
    }
  }

  rc = fs_close(&file);
  if (rc < 0) {
    printk("couldn't close file: %d\n", rc);
    return 1;
  }
  return 0;
}
```

nucleo_h723zg.overlay
```
/ {
	fstab {
		compatible = "zephyr,fstab";
		lfs1: lfs1 {
			compatible = "zephyr,fstab,littlefs";
            /* Flashword size of STM32H732ZG internal flash is 32B */
			read-size = <32>;
			prog-size = <32>;
			cache-size = <256>;
			lookahead-size = <32>;
			block-cycles = <512>;
			partition = <&storage_partition>;
			mount-point = "/seq";
    	};
	};
};

&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		/* Use second half of flash for the filesystem. */
		storage_partition: partition@80000 {
			label = "storage";
			reg = <0x80000 DT_SIZE_K(512)>;
		};
	};
};
```

### Output
The output I get from the test setup:
```
*** Booting Zephyr OS build zephyr-v3.3.0 ***                                  
Area 0 at 0x80000 on flash-controller@52002000 for 524288 bytes                
E: Erasing flash area ... 0
I: LittleFS version 2.5, disk version 2.0                                      
I: FS at flash-controller@52002000:0x80000 is 4 0x20000-byte blocks with 512 cycle
I: sizes: rd 32 ; pr 32 ; ca 256 ; la 32                                       
E: WEST_TOPDIR/modules/fs/littlefs/lfs.c:1234: Corrupted dir pair at {0x0, 0x1} 
W: can't mount (LFS -84); formatting
I: /seq mounted                        
/seq mount: 0                          
35 (cur) -1                            
45 (cur) 35
Writing file                            
overrun, got: 45 exp: 36
46 (cur) 45                            
47 (cur) 46
48 (cur) 47
49 (cur) 48
...
```

I send new sequence number every 100ms and the overrun in the output happens during the file write.

**Comment by nordicjm** [comment]
You cannot execute from flash whilst erasing/programming flash, the solution is to either use hardware flow control or use a protocol like MCUmgr that sends a command and waits for a response before continuing

**Comment by sobkulir** [comment, marked as answer]
Thanks @nordicjm, you reassured me that I'm on the right path so I dug a bit deeper. This is the sentence in the manual[1] confirming what you've said about only a single read/write/erase operation being able to be performed at a time (although it's not super clearly written so should be taken with a grain of salt).
![image](https://github.com/zephyrproject-rtos/zephyr/assets/14258647/87d44c51-414d-4799-9dba-c2f9ba106f53)

What I also didn't know is that Flash memory can only be erased in sectors (128kB), but programmed in words (32B). Numbers are for stm32h723zg internal flash.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/14258647/f9e9da86-f384-4e8c-80c8-e89f9e13d1a6)

So what is likely happening is that 
1) LittleFS tries to change the size of the file in the metadata -- it needs to flip some 0s back to 1s, but for that it needs to erase the whole sector.
2) Erasing the sector takes a considerable amount of time.
3) In the meantime, interrupts for the UART arrive, but the interrupt vector table (IVT) and the interrupt handler code are all in the flash. The code cannot be read from the flash so interrupts are not handled.

Except for the solutions mentioned by @nordicjm, one could put the IVT and the handler into RAM to work around this. Well in any case, using the internal flash of this MCU for a dynamic filesystem is probably not a good idea, as it can only sustain 10000 write/erase cycles[2].
![image](https://github.com/zephyrproject-rtos/zephyr/assets/14258647/d3acf519-1188-4968-872c-36b43d18af28)
 

[1] [RM0468 Rev 3](https://www.st.com/resource/en/reference_manual/rm0468-stm32h723733-stm32h725735-and-stm32h730-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)
[2] [DS13313 Rev3](https://www.st.com/resource/en/datasheet/stm32h723ve.pdf)

**Reply by thedjnK** [reply]
Yes, I did experiment with this with other MCUs trying to allow for MCUmgr to perform an async erase but after hitting limitations whereby UART data was dropped if an erase was in progress, gave up on it entirely. Some MCUs (including STM32) have dual access to flash, so *in theory* they can execute whilst an erase is in progress (not sure if zephyr would support that, I would assume it would not) but in general most MCUs have single access to flash.
I would recommend MCUmgr for this as it has a file system group which allows for uploading to/downloading from filesystems, and there's other groups for things like FOTA updates which can be enabled simply by enabling a Kconfig.
